### PR TITLE
Clone for IncrementalHash and KMAC

### LINKED ISSUE #103170: [API Proposal]: Clone for IncrementalHash and KMAC
### Background and motivation

There has been some interest in making `IncrementalHash` clone-able, in the sense that cloning it duplicates the state of the hash. That was originally proposed in https://github.com/dotnet/runtime/issues/1968, however at the time it was believed that a "give me the current hash" was sufficient, and we could add `Clone` later if it still seemed useful.

Some feedback has shown that it would be useful to be able to snapshot and rewind an `IncrementalHash`, such as accumulating file transfer chunks, and if a chunk fails, go back to the last known good chunk and re-download the chunk.

`Shake128` and `Shake256` also introduced a `Clone`, so we have precedent for the name and API.

KMAC would be the only hash/mac primitive that does not have Clone, so we should consider adding it for consistency.

### API Proposal

```csharp
namespace System.Security.Cryptography;

public partial class IncrementalHash {
    public IncrementalHash Clone();
}

public partial class Kmac128 {
    public Kmac128 Clone();
}

public partial class KmacXof128 {
    public KmacXof128 Clone();
}

public partial class Kmac256 {
    public Kmac256 Clone();
}

public partial class KmacXof256 {
    public KmacXof256 Clone();
}
```

### API Usage

```C#
using IncrementalHash ih = new(HashAlgorithmName.SHA256);
ih.AppendData("potato"u8);

using IncrementalHash clone = ih.Clone();
ih.AppendData("tomatoes"u8);

// clone does not have tomatoes appended to it.
```

### Alternative Designs

I had considered if the `HashAlgorithm` derived types should implement a `Clone` but decided against it for a few reasons.

1. It would need to be `virtual` and by default it would need to throw since we don't know how to duplicate arbitrary hash algorithms.
2. Personally, I consider `HashAlgorithm` "done" and we should not innovate in its API surface. API innovation should either be in the static one-shots, `IncrementalHash`, or non-HashAlgorithm derived types that are sealed, like `Shake128` and `Kmac128`, etc. If developers want a `Clone` for their own `HashAlgorithm` derived types, they are free to add them.

### Risks

_No response_

### PR DESCRIPTION
**Pull request by vcsjones**
Closes #103170

### CODE REVIEW

File: src/native/libs/System.Security.Cryptography.Native/entrypoints.c [hunk 1]
```diff
@@ -145,11 +145,12 @@ static const Entry s_cryptoNative[] =
     DllImportEntry(CryptoNative_EvpDigestSqueeze)
     DllImportEntry(CryptoNative_EvpDigestUpdate)
     DllImportEntry(CryptoNative_EvpDigestXOFOneShot)
-    DllImportEntry(CryptoNative_EvpMacFetch)
-    DllImportEntry(CryptoNative_EvpMacFree)
+    DllImportEntry(CryptoNative_EvpMacCtxDup)
```
Comment (vcsjones) [inline src/native/libs/System.Security.Cryptography.Native/entrypoints.c:148]: Some of this was just sorting. The only meaningful change here was adding MacCtxDup and HmacCopy.

File: src/native/libs/System.Security.Cryptography.Native/pal_hmac.h [hunk 2]
```diff
@@ -81,3 +81,10 @@ PALEXPORT int32_t CryptoNative_HmacOneShot(const EVP_MD* type,
                                            int32_t sourceSize,
                                            uint8_t* md,
                                            int32_t* mdSize);
+
+
+/**
```
Comment (bartonjs) [inline src/native/libs/System.Security.Cryptography.Native/pal_hmac.h:86]:
```suggestion

/**
```

File: src/native/libs/System.Security.Cryptography.Native.Apple/pal_hmac.h [hunk 3]
```diff
@@ -65,3 +65,11 @@ PALEXPORT int32_t AppleCryptoNative_HmacOneShot(PAL_HashAlgorithm algorithm,
                                                 uint8_t* pOutput,
                                                 int32_t cbOutput,
                                                 int32_t* pcbDigest);
+
+
+/*
```
Comment (bartonjs) [inline src/native/libs/System.Security.Cryptography.Native.Apple/pal_hmac.h:70]:
```suggestion

/*
```

**Review by bartonjs — APPROVED** [review]
Just a couple of superfluous blank lines.  If removing them would put making the preview 6 cutoff (Tuesday afternoon, PT) at risk, merge with them here.  If you have the time, I'd prefer them tidied up.

**Comment by vcsjones** [comment]
Since the only change from the last run was white space changes, I am not going to run the Android / iOS / Browser legs again. And because they did not fail in S.S.Cryptography, merging.

**Pull request merged**